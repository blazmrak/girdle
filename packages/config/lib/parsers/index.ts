import { StringVar } from './string'
import { NumberVar } from './number'
import { BooleanVar } from './boolean'
import { ArrayVar } from './array'
import { EnumVar } from './enum'

export type Value = string | number | boolean
export type ValueStr = 'string' | 'number' | 'boolean'
export type ExtractValue<T extends ValueStr> = T extends 'string'
  ? string
  : T extends 'number'
  ? number
  : boolean

export type ParseError = { error: string }
export type ParseParams = { path: string; name: string }

export type BaseVarOpts = {
  type: string
}

export class Var {
  static string() {
    return new StringVar()
  }

  static number() {
    return new NumberVar()
  }

  static boolean() {
    return new BooleanVar()
  }

  static array<T extends ValueStr>(contains: T) {
    return new ArrayVar({ type: 'array', contains })
  }

  static enum<T extends { [key: string]: string } | readonly string[]>(output: T): EnumVar<T> {
    return new EnumVar({ type: 'enum', enum: output })
  }
}
