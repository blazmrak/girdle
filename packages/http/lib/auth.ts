import {
  applyDecorators,
  CanActivate,
  createParamDecorator,
  ExecutionContext,
  Injectable,
  UnauthorizedException,
  UseGuards,
} from '@nestjs/common'
import * as jwt from 'jsonwebtoken'
import {
  ApiBearerAuth,
  ApiExtension,
  ApiForbiddenResponse,
  ApiUnauthorizedResponse,
} from '@nestjs/swagger'

type AuthGuardOpts = {
  secret: string
}

@Injectable()
export class AuthGuard implements CanActivate {
  constructor(private opts: AuthGuardOpts) {}
  async canActivate(context: ExecutionContext): Promise<boolean> {
    const request = context.switchToHttp().getRequest()
    const [type, token] = request.headers.authorization?.split(' ') ?? [undefined, undefined]

    if (type === 'Bearer') {
      try {
        request['user'] = jwt.verify(token, this.opts.secret)
      } catch {}
    }

    return true
  }
}

export function Protected(...roles: string[]) {
  const ext = roles?.map(r => r.toUpperCase())?.join(', ') || 'ALL'
  return applyDecorators(
    UseGuards(new RolesGuard(roles)),
    ApiBearerAuth(),
    ApiExtension('x-roles', ext),
    ApiForbiddenResponse({ description: 'Insufficient privileges' }),
    ApiUnauthorizedResponse({ description: 'Missing or invalid token' }),
  )
}

export const Token = createParamDecorator(
  (data = { required: true }, context: ExecutionContext) => {
    const request = context.switchToHttp().getRequest()
    const user = request['user']

    if (!user && data.required) {
      throw new UnauthorizedException()
    }

    return user
  },
)

@Injectable()
export class RolesGuard implements CanActivate {
  constructor(private readonly roles: string[]) {}

  async canActivate(context: ExecutionContext): Promise<boolean> {
    const request = context.switchToHttp().getRequest()
    const roles = <string[] | undefined>request?.user?.roles

    if (!roles) {
      throw new UnauthorizedException()
    }

    if (this.roles.length === 0) {
      return true
    }

    return this.roles.some(r => roles.some(u => u === r))
  }
}
