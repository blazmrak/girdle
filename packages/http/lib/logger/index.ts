import { LoggerModuleAsyncParams, Params } from 'nestjs-pino'
import { LogContent, LogFormat, LogLevel } from './types'

export * from './types'

function initTransport(
  environment: string,
  level: LogLevel,
  format: LogFormat,
  content: LogContent[],
): any {
  if (format == LogFormat.PRETTY) {
    let messageFormat = ''
    if (content.includes(LogContent.META)) {
      messageFormat += '{req.meta.id} - {req.meta.method} {req.meta.url} '
    }
    if (content.includes(LogContent.USER)) {
      messageFormat += '{req.user.roles} '
    }
    messageFormat += '{res.statusCode} ({responseTime}ms): {msg}'

    return {
      target: 'pino-pretty',
      options: {
        colorize: true,
        timestampKey: null,
        messageFormat,
        hideObject: level !== LogLevel.TRACE,
        sync: environment === 'test',
      },
    }
  }
}

function extractUser(req: any) {
  const auth = req?.headers?.authorization
  if (typeof auth === 'string' && auth.startsWith('Bearer')) {
    const [, token] = auth.split(' ')
    const [, user] = token.split('.')
    const parsed = Buffer.from(user, 'base64').toString()
    return JSON.parse(parsed)
  }
}

function requestSerializer(logContent: LogContent[]) {
  const serializers: ((req: any, log: any) => any)[] = []
  if (logContent.includes(LogContent.META)) {
    serializers.push((req: any, log: any) => {
      log.meta ??= {}
      log.meta.id = req.id
      log.meta.url = req.url
      log.meta.method = req.method
    })
  }
  if (logContent.includes(LogContent.PARAMS)) {
    serializers.push((req: any, log: any) => {
      log.meta ??= {}
      log.meta.query = req.query
      log.meta.params = req.params
    })
  }
  if (logContent.includes(LogContent.USER)) {
    serializers.push((req: any, log: any) => {
      const user = extractUser(req)
      log.user = {
        jwt: user,
        name: user?.sub ?? '/',
        roles: user?.roles.join(', ') ?? '/',
      }
    })
  }
  if (logContent.includes(LogContent.HEADERS)) {
    serializers.push((req: any, log: any) => {
      log.headers = req.headers
    })
  }
  if (logContent.includes(LogContent.REQ_BODY)) {
    serializers.push((req: any, log: any) => {
      log.body = req.body
    })
  }

  return (req: any) => {
    const log = {}
    for (const serializer of serializers) {
      serializer(req, log)
    }

    return log
  }
}

function responseSerializer(logContent: LogContent[]) {
  if (logContent.includes(LogContent.RES_BODY)) {
    return (res: any) => {
      return {
        statusCode: res.statusCode,
        body: res.responseBody,
      }
    }
  } else {
    return (res: any) => {
      return {
        statusCode: res.statusCode,
      }
    }
  }
}

export type ZarjaLoggerConfig = {
  transport: {
    env: string
    level: LogLevel
    content: LogContent[]
    format: LogFormat
  }
  cfg?: Omit<Params, 'pinoHttp'>
}

export function createLoggerConfig(
  {
    transport: {
      level = LogLevel.INFO,
      content = [LogContent.META, LogContent.USER],
      format = LogFormat.RAW,
      env = 'production',
    },
    cfg,
  }: ZarjaLoggerConfig = {
    transport: {
      level: LogLevel.INFO,
      content: [LogContent.META, LogContent.USER],
      format: LogFormat.RAW,
      env: 'production',
    },
  },
): LoggerModuleAsyncParams {
  return {
    useFactory: (): Params => {
      return {
        pinoHttp: {
          mixin(_context, level) {
            let label
            switch (level) {
              case 10:
                label = 'trace'
                break
              case 20:
                label = 'debug'
                break
              case 30:
                label = 'info'
                break
              case 40:
                label = 'warn'
                break
              case 50:
                label = 'error'
                break
            }
            return { 'level-label': label }
          },
          transport: initTransport(env, level, format, content),
          customLogLevel: (req, res) => {
            if (res.statusCode > 500) {
              return 'error'
            } else {
              return 'info'
            }
          },
          customSuccessMessage: (req, res) => {
            if (400 <= res.statusCode && res.statusCode < 500 && 'responseBody' in res) {
              return JSON.stringify(res.responseBody)
            } else {
              return 'request success'
            }
          },
          wrapSerializers: false,
          level: level,
          serializers: {
            req: requestSerializer(content),
            res: responseSerializer(content),
          },
        },
        ...cfg,
      }
    },
  }
}
