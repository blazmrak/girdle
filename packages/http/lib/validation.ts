import { applyDecorators } from '@nestjs/common'
import {
  IsEnum as CVIsEnum,
  IsNumber,
  IsOptional as CVIsOptional,
  IsString,
  Max as CVMax,
  MaxLength as CVMaxLength,
  Min as CVMin,
  MinLength as CVMinLength,
  ValidationOptions,
} from 'class-validator'
import { ApiProperty } from '@nestjs/swagger'
import { Transform } from 'class-transformer'

export const Max = (maximum: number) =>
  applyDecorators(IsNumber(), CVMax(maximum), ApiProperty({ maximum }))

export const Min = (minimum: number) =>
  applyDecorators(IsNumber(), CVMin(minimum), ApiProperty({ minimum }))

export const MaxLength = (maxLength: number) =>
  applyDecorators(IsString(), CVMaxLength(maxLength), ApiProperty({ maxLength }))

export const MinLength = (minLength: number) =>
  applyDecorators(IsString(), CVMinLength(minLength), ApiProperty({ minLength }))

export const IsEnum = (value: object, opts?: ValidationOptions) =>
  applyDecorators(CVIsEnum(value, opts), ApiProperty({ enum: value, isArray: opts?.each }))

export const IsOptional = () => applyDecorators(CVIsOptional(), ApiProperty({ required: false }))

export const Default = (defaultValue: unknown) => {
  if (defaultValue instanceof Function) {
    return applyDecorators(
      Transform((value: unknown) => value ?? defaultValue()),
      ApiProperty({ default: defaultValue() }),
    )
  }

  return applyDecorators(
    Transform((value: unknown) => value ?? defaultValue),
    ApiProperty({ default: defaultValue }),
  )
}
