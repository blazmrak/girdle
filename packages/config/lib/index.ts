import { BaseVarOpts, ParseError, Value, Var } from './parsers'
import { BaseVar, TransformationContext } from './parsers/base'
import * as dotenv from 'dotenv'
import * as path from 'path'
import * as fs from 'fs'
export { Var } from './parsers'

export type ConfigDefinitionRec<V extends BaseVarOpts, T extends Value | Value[] | undefined> = {
  [key: string]: ConfigDefinitionRec<V, T> | BaseVar<V, T>
}

export type ConfigDefinition<
  V extends BaseVarOpts,
  T extends Value | Value[] | undefined,
> = ConfigDefinitionRec<V, T> & { env?: BaseVar<V, T> }

export type ExtractConfig<C extends ConfigDefinition<any, any>> = {
  [key in keyof C]: C[key] extends BaseVar<any, any>
    ? C[key]['output']
    : C[key] extends ConfigDefinitionRec<any, any>
    ? ExtractConfig<C[key]>
    : never
}

export type Config<C extends ConfigDefinition<any, any>> = ExtractConfig<C> & {
  env: C['env'] extends BaseVar<any, any>
    ? C['env']['output']
    : 'local' | 'test' | 'staging' | 'production'
}

type InitOpts = {
  env?: Record<string, string | undefined>
  file?: string
  directory?: string
  fromFile?: boolean
}

type ParsedResult = { [key: string]: Value | Value[] | undefined | ParsedResult }
type ParsedNode =
  | {
      result: ParsedResult
    }
  | {
      errors: ParseError[]
    }

function parseEnvironment<T extends ConfigDefinitionRec<BaseVarOpts, Value | Value[]>>(
  template: T,
  env: Record<string, string | undefined>,
  path = '',
): ParsedNode {
  const errors: ParseError[] = []

  const result = Object.entries(template).reduce((acc, [key, value]) => {
    const newPath = path ? path + '.' + key : key
    const name = path ? (path + '_' + key).toUpperCase() : key.toUpperCase()
    if (value instanceof BaseVar) {
      const res = value.parse(env, { path: newPath, name })
      if (typeof res === 'object' && !Array.isArray(res)) {
        errors.push(res)
      } else {
        acc[key] = res
      }
    } else {
      const res = parseEnvironment(value, env, newPath)
      if ('errors' in res) {
        errors.push(...res.errors)
      } else {
        acc[key] = res.result
      }
    }
    return acc
  }, {} as ParsedResult)

  if (errors.length > 0) {
    return { errors }
  } else {
    return { result }
  }
}

function runTransformations(
  config: any,
  template: ConfigDefinitionRec<any, any>,
  env: string,
  path = '',
): ParsedNode {
  const errors: ParseError[] = []

  const result = Object.entries(template).reduce((acc, [key, value]) => {
    const newPath = path ? path + '.' + key : key
    const name = path ? (path + '_' + key).toUpperCase() : key.toUpperCase()
    if (value instanceof BaseVar) {
      const res = value.runTransformations(
        config[key],
        new TransformationContext({ env, path: newPath, name }),
      )
      if (typeof res === 'object' && !Array.isArray(res)) {
        errors.push(res)
      } else {
        acc[key] = res
      }
    } else {
      const res = runTransformations(config[key], value, env, newPath)
      if ('errors' in res) {
        errors.push(...res.errors)
      } else {
        acc[key] = res.result
      }
    }
    return acc
  }, config)

  if (errors.length > 0) {
    return { errors }
  } else {
    return { result }
  }
}

type Variable = { name: string; path: string; type: string; required: boolean; default: any }

function extractVariables(
  template: ConfigDefinitionRec<any, any>,
  env: string,
  path = '',
): Variable[] {
  return Object.entries(template).reduce((acc, [key, value]) => {
    const newPath = path ? path + '.' + key : key
    const name = path ? (path + '_' + key).toUpperCase() : key.toUpperCase()
    if (value instanceof BaseVar) {
      const actualName = value.getName(name)
      let required = !value['_optional'] && value['_default'] === undefined
      let def = required
        ? undefined
        : value.runTransformations(
            value['_default'],
            new TransformationContext({ env, path: newPath, name: name }),
          )
      acc.push({
        name: actualName,
        path: newPath,
        type: value['_opts']['type'],
        required,
        default: def,
      })
    } else {
      const res = extractVariables(value, env, newPath)
      acc.push(...res)
    }
    return acc
  }, [] as Variable[])
}

class EnvironmentInitializationFailed {
  constructor(public errors: ParseError[]) {}
}

export function getVariables<T extends ConfigDefinition<any, any>>(
  template: T,
  environment?: string,
): Variable[] {
  if (environment) {
    template.env = Var.enum(environment).name(template.env?.getName('ENV') ?? 'NODE_ENV')
  }

  template.env ??= Var.enum(['local', 'test', 'staging', 'production'])
    .name('NODE_ENV')
    .default('local')
  let env = template.env['_default']
  env = template.env.runTransformations(
    env,
    new TransformationContext({ env, path: 'env', name: template.env.getName('ENV') }),
  )

  return extractVariables(template, env)
}

export function renderEnvFileContent(variables: Variable[]) {
  return variables
    .map(v => {
      if (!v.required && v.default == null) {
        return `#${v.name}=`
      } else if (v.default != null) {
        return `#${v.name}=${v.default}`
      } else {
        return `${v.name}=`
      }
    })
    .join('\n')
}

function readFile(directory: string, file: string) {
  const filePath = path.resolve(path.join(directory, file))
  let content
  try {
    content = fs.readFileSync(filePath)
  } catch {
    console.warn(`error reading .env file at ${filePath}, defaulting to process.env`)
    return process.env
  }

  try {
    return dotenv.parse(content)
  } catch {
    console.warn(`failed to parse '${file}', defaulting to process.env`)
    return process.env
  }
}

export function initializeEnvironment<T extends ConfigDefinition<any, any>>(
  template: T,
  { env, file = '.env.local', directory = './', fromFile = true }: InitOpts = {},
): Config<T> {
  if (!env && !fromFile) {
    env = process.env
  } else if (!env) {
    env = readFile(directory, file)
  }

  template.env ??= Var.enum(['local', 'test', 'staging', 'production'])
    .name('NODE_ENV')
    .default('local')
  const res = parseEnvironment(template, env)

  if ('errors' in res) throw new EnvironmentInitializationFailed(res.errors)

  const result = runTransformations(res.result, template, (res.result as any).env)

  if ('errors' in result) throw new EnvironmentInitializationFailed(result.errors)

  return result.result as any
}
