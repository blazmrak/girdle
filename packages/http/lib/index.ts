import { INestApplication, ValidationPipe } from '@nestjs/common'
import { ClassSerializerInterceptor, ExcludeNullInterceptor, PaginationDto } from './serialization'
import { Reflector } from '@nestjs/core'
import helmet, { HelmetOptions } from 'helmet'
import { AuthGuard } from './auth'
import { DocumentBuilder, SwaggerModule } from '@nestjs/swagger'

export * from './pagination'
export * from './validation'
export * from './serialization'

type PackerHttpOpts = {
  helmet: HelmetOptions
  serialization: {
    serializeResponses: boolean
    excludeNullFromResponse: boolean
  }
  jwt?: {
    enabled: boolean
    secret: string
  }
  swagger?: {
    enabled: boolean
    version: string
    title: string
    document?: {
      jsonPath?: string
      yamlPath?: string
    }
    gui: {
      path: string
    }
  }
}

export class ZarjaHttp {
  static setup(app: INestApplication, opts: PackerHttpOpts) {
    app.use(helmet(opts.helmet))

    if (opts.jwt?.enabled) app.useGlobalGuards(new AuthGuard(opts.jwt))

    app.useGlobalPipes(new ValidationPipe({ whitelist: true }))

    if (opts.serialization.serializeResponses)
      app.useGlobalInterceptors(new ClassSerializerInterceptor(new Reflector()))

    if (opts.serialization.excludeNullFromResponse)
      app.useGlobalInterceptors(new ExcludeNullInterceptor())

    if (opts.swagger?.enabled) {
      const config = new DocumentBuilder()
        .setTitle(opts.swagger.title)
        .setVersion(opts.swagger.version)
        .addBearerAuth()
        .build()

      const document = SwaggerModule.createDocument(app, config, { extraModels: [PaginationDto] })
      SwaggerModule.setup(opts.swagger.gui.path, app, document, {
        jsonDocumentUrl: opts.swagger.document?.jsonPath,
        yamlDocumentUrl: opts.swagger.document?.yamlPath,
        swaggerOptions: <any>{
          docExpansion: 'list',
          filter: true,
          showExtensions: true,
          showCommonExtensions: true,
          defaultModelExpandDepth: 100,
          defaultModelsExpandDepth: 100,
        },
      })
    }
  }
}
