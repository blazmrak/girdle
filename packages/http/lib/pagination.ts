import { IsOptional, Max, Min } from './validation'
import { Public } from './serialization'
import { IsString } from 'class-validator'

export interface Pagination<T> {
  data: T[]
  page: number
  length: number
  maxPage: number
}

export class PaginationParams {
  @IsOptional()
  @Min(1)
  // @Default(1)
  @Public()
  page?: number

  @IsOptional()
  @Max(100)
  // @Default(10)
  @Public()
  length?: number
}

export class SearchParams extends PaginationParams {
  @IsOptional()
  @IsString()
  @Public()
  search?: string
}
