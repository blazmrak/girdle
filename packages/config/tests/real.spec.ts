import { getVariables, initializeEnvironment, renderEnvFileContent, Var } from '../lib'
import { expectTypeOf } from 'expect-type'
import { extractError } from './helpers/assertions'

enum LogLevel {
  ERROR = 'error',
  INFO = 'info',
}

describe('real tests', function () {
  let configSchema = {
    port: Var.number().default(8080),
    jwt: {
      signingSecret: Var.string().name('JWT_SECRET'),
      expiresIn: Var.number()
        .name('JWT_EXPIRES_IN')
        .default(1000 * 60 * 5),
    },
    log: {
      level: Var.enum(LogLevel).default(LogLevel.INFO),
      format: Var.enum(['raw', 'pretty'] as const).default('raw'),
      content: Var.array(['meta', 'user', 'body'])
        .default(['meta', 'user'])
        .transform(['staging'], arg => arg.filter(i => i !== 'body')),
    },
    security: {
      passwordRounds: Var.number().name('PASSWORD_ROUNDS').default(12),
    },
    db: {
      url: Var.string(),
      type: Var.enum(['postgres'] as const).default('postgres'),
      logging: Var.boolean()
        .default(false)
        .transform(['staging', 'production'], () => false),
      recreate: Var.boolean().default(false),
    },
    swagger: Var.boolean().default(false),
  }

  it('simple configuration', () => {
    const env = { JWT_SECRET: 'secret', DB_URL: 'db.url' }

    const configuration = initializeEnvironment(configSchema, { env })

    expect(configuration).toMatchSnapshot()
    expectTypeOf(configuration).toEqualTypeOf<{
      env: 'local' | 'test' | 'staging' | 'production'
      port: number
      jwt: {
        signingSecret: string
        expiresIn: number
      }
      log: {
        level: LogLevel
        format: 'raw' | 'pretty'
        content: string[]
      }
      security: {
        passwordRounds: number
      }
      db: {
        url: string
        type: 'postgres'
        logging: boolean
        recreate: boolean
      }
      swagger: boolean
    }>()
  })

  it('reports errors', () => {
    const env = {}

    const configuration = extractError(() => initializeEnvironment(configSchema, { env }))

    expect(configuration).toMatchSnapshot()
  })

  it('extracts all environment variables', () => {
    const variables = getVariables(configSchema)

    expect(variables).toMatchSnapshot()
    expect(variables.filter(v => v.required)).toMatchSnapshot()
  })

  it('renders env file', () => {
    const variables = getVariables(configSchema)

    const fileContent = renderEnvFileContent(variables)

    expect(fileContent).toMatchSnapshot()
  })
})
