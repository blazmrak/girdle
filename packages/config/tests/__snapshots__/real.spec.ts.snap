// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`real tests extracts all environment variables 1`] = `
[
  {
    "default": 8080,
    "name": "PORT",
    "path": "port",
    "required": false,
    "type": "number",
  },
  {
    "default": undefined,
    "name": "JWT_SECRET",
    "path": "jwt.signingSecret",
    "required": true,
    "type": "string",
  },
  {
    "default": 300000,
    "name": "JWT_EXPIRES_IN",
    "path": "jwt.expiresIn",
    "required": false,
    "type": "number",
  },
  {
    "default": "info",
    "name": "LOG_LEVEL",
    "path": "log.level",
    "required": false,
    "type": "enum",
  },
  {
    "default": "raw",
    "name": "LOG_FORMAT",
    "path": "log.format",
    "required": false,
    "type": "enum",
  },
  {
    "default": [
      "meta",
      "user",
    ],
    "name": "LOG_CONTENT",
    "path": "log.content",
    "required": false,
    "type": "array",
  },
  {
    "default": 12,
    "name": "PASSWORD_ROUNDS",
    "path": "security.passwordRounds",
    "required": false,
    "type": "number",
  },
  {
    "default": undefined,
    "name": "DB_URL",
    "path": "db.url",
    "required": true,
    "type": "string",
  },
  {
    "default": "postgres",
    "name": "DB_TYPE",
    "path": "db.type",
    "required": false,
    "type": "enum",
  },
  {
    "default": false,
    "name": "DB_LOGGING",
    "path": "db.logging",
    "required": false,
    "type": "boolean",
  },
  {
    "default": false,
    "name": "DB_RECREATE",
    "path": "db.recreate",
    "required": false,
    "type": "boolean",
  },
  {
    "default": false,
    "name": "SWAGGER",
    "path": "swagger",
    "required": false,
    "type": "boolean",
  },
  {
    "default": "local",
    "name": "NODE_ENV",
    "path": "env",
    "required": false,
    "type": "enum",
  },
]
`;

exports[`real tests extracts all environment variables 2`] = `
[
  {
    "default": undefined,
    "name": "JWT_SECRET",
    "path": "jwt.signingSecret",
    "required": true,
    "type": "string",
  },
  {
    "default": undefined,
    "name": "DB_URL",
    "path": "db.url",
    "required": true,
    "type": "string",
  },
]
`;

exports[`real tests reports errors 1`] = `
EnvironmentInitializationFailed {
  "errors": [
    {
      "error": "Value of 'jwt.signingSecret' (JWT_SECRET) should be a string",
    },
    {
      "error": "Value of 'db.url' (DB_URL) should be a string",
    },
  ],
}
`;

exports[`real tests simple configuration 1`] = `
{
  "db": {
    "logging": false,
    "recreate": false,
    "type": "postgres",
    "url": "db.url",
  },
  "env": "local",
  "jwt": {
    "expiresIn": 300000,
    "signingSecret": "secret",
  },
  "log": {
    "content": [
      "meta",
      "user",
    ],
    "format": "raw",
    "level": "info",
  },
  "port": 8080,
  "security": {
    "passwordRounds": 12,
  },
  "swagger": false,
}
`;
